rpc/rpc.go

```diff
@@ -1256,245 +1249,234 @@ func (c *Conn) handleRelease(ctx context.Context, id exportID, count uint32) err
+			defer release()
 
-func (c *Conn) handleUnknownMessage(ctx context.Context, recv rpccp.Message) error {
-	c.report(failedf("unknown message type %v from remote", recv.Which()))
-	var err error
-	syncutil.With(&c.mu, func() {
-		err = c.sendMessage(ctx, func(msg rpccp.Message) error {
-			return msg.SetUnimplemented(recv)
+			if m, err = m.NewUnimplemented(); err == nil {
+				err = m.SetDisembargo(d)
+			}
+
+			if err != nil {
+				err = rpcerr.Annotate(err, "incoming disembargo: send unimplemented")
+			}
+
+			return
 		})
-	})
-	if err != nil {
-		c.report(annotate(err, "send unimplemented"))
 	}
+
 	return nil
 }
 
 // startTask increments c.tasks if c is not shutting down.
 // It returns whether c.tasks was incremented.
 //
 // The caller must be holding onto c.mu.
-func (c *Conn) startTask() bool {
-	select {
-	case <-c.bgctx.Done():
-		return false
-	default:
+func (c *Conn) startTask() (ok bool) {
+	if c.bgctx.Err() == nil {
 		c.tasks.Add(1)
-		return true
+		ok = true
 	}
+
+	return
 }
 
-// sendMessage creates a new message on the Sender, calls f to build it,
-// and sends it if f does not return an error.  When f returns, the
-// message must have a nil CapTable.  The caller must be holding onto
-// c.mu.  While f is being called, it will be holding onto the sender
-// lock, but not c.mu.
-func (c *Conn) sendMessage(ctx context.Context, f func(msg rpccp.Message) error) error {
-	if err := c.tryLockSender(ctx); err != nil {
-		return err
-	}
-	c.mu.Unlock()
-	msg, send, release, err := c.transport.NewMessage(ctx)
-	if err != nil {
-		c.mu.Lock()
-		c.unlockSender()
-		return failedf("create message: %w", err)
-	}
-	if err := f(msg); err != nil {
-		release()
-		c.mu.Lock()
-		c.unlockSender()
-		return failedf("build message: %w", err)
-	}
-	err = send()
-	release()
-	c.mu.Lock()
-	c.unlockSender()
-	if err != nil {
-		return failedf("send message: %w", err)
+func (c *Conn) sendMessage(ctx context.Context, f func(rpccp.Message) error) (err error) {
+	cherr, release := c.sender.Send(c.newSender(ctx, f))
+	defer release()
+
+	select {
+	case err = <-cherr:
+	case <-ctx.Done():
+		err = ctx.Err()
+	case <-c.bgctx.Done():
+		err = c.bgctx.Err()
 	}
-	return nil
+
+	return
 }
 
-// tryLockSender attempts to acquire the sender lock, returning an error
-// if either the Context is Done or c starts shutdown before the lock
-// can be acquired.  The caller must be holding c.mu.
-func (c *Conn) tryLockSender(ctx context.Context) error {
-	for {
-		select {
-		case <-c.bgctx.Done():
-			return ExcClosed
-		default:
-		}
-		s := c.sendCond
-		if s == nil {
-			break
-		}
-		c.mu.Unlock()
-		select {
-		case <-s:
-		case <-ctx.Done():
-			c.mu.Lock()
-			return ctx.Err()
-		case <-c.bgctx.Done():
-			c.mu.Lock()
-			return ExcClosed
-		}
-		c.mu.Lock()
+func (c *Conn) sendMessageAsync(ctx context.Context, f func(rpccp.Message) error) {
+	cherr, release := c.sender.Send(c.newSender(ctx, f))
+	defer release()
+
+	select {
```

Comment:
Wait, am I misunderstanding? It looks like this *also* blocks, i.e. is not actually async. How is this supposed to work?

Up until reading this part I was assuming that this function was non-blocking, which was the whole point of this pr, but it looks like it just does the same thing as sendMessage, except it calls ReportError instead of returning the error.